File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.py
def solution(str1, str2):
    intersection = 0
    union = 0

    multiset1 = to_multiset(str1)
    multiset2 = to_multiset(str2)

    for i in range(26):
        for j in range(26):
            intersection += min(multiset1[i][j], multiset2[i][j])
            union += max(multiset1[i][j], multiset2[i][j])

    if union == 0:
        return 65536
    else:
        return int(intersection / union * 65536)

def to_multiset(string:str):
    multiset = [[0]*26 for __ in range(26)]

    for i in range(len(string)-1):
        if string[i].isalpha() and string[i+1].isalpha():
            multiset[ord(string[i].upper())-65][ord(string[i+1].upper())-65] += 1
    
    return multiset

File: 프로그래머스/2/17677. ［1차］ 뉴스 클러스터링/［1차］ 뉴스 클러스터링.java
class Solution {
    public int solution(String str1, String str2) {
        float intersection = 0;
        float union = 0;

        int[][] multiset1 = toMultiset(str1);
        int[][] multiset2 = toMultiset(str2);
        for (int i = 0; i < 26; i++) {
            for (int j = 0; j < 26; j++) {
                intersection += Math.min(multiset1[i][j], multiset2[i][j]);
                union += Math.max(multiset1[i][j], multiset2[i][j]);
            }
        }

        return union == 0
                ? 65536
                : (int) (intersection / union * 65536);
    }

    private int[][] toMultiset(String str) {
        int[][] multiset = new int[26][26];

        char[] arr = str.toUpperCase().toCharArray();
        for (int i = 0; i < arr.length - 1; i++) {
            char c1 = arr[i];
            char c2 = arr[i + 1];
            if (Character.isAlphabetic(c1) && Character.isAlphabetic(c2)) {
                multiset[c1 - 65][c2 - 65]++;
            }
        }

        return multiset;
    }
}